src/injectors-cache.ts

```diff
@@ -0,0 +1,157 @@
+ *     fetching the Brotli pre-compressed version of the URL.
+ * @returns an already injected Response.
+ */
+export async function getCacheFor(
+  url: string,
+  key: string,
+  cf: IncomingCloudflarePropertiesExtended
+): Promise<Response | undefined> {
+  const cache = await openInjectorsCache();
+  return supportsBrotli(cf)
+    ? cache.match(cacheKeyBrotli(url, key))
+    : cache.match(cacheKey(url, key));
+}
+
+/**
+ * Saves an injected Response to the cache twice, for raw and Brotli-compressed.
+ *
+ * @param response - raw Response to save to cache.
+ * @param url - base file URL that has dynamic content injected into it.
+ * @param key - unique identifiers for this specific injection.
+ */
+async function saveCache(
+  response: Response,
+  url: string,
+  key: string
+): Promise<void> {
+  console.log('Caching', url, 'with cache key', key);
+  if (!response.body) {
+    throw new Error('Response has no body');
+  }
+
+  // Replace whatever cache-control header that we respond with, with a 1 year
+  // cache-control. Responses to user requests should replace this header again
+  // with the relevant cache-control before serving.
+  response.headers.set(HeaderKeys.CACHE_CONTROL, CacheControl.STATIC_RTV_FILE);
+
+  // We need to read the body of the response twice: once for saving the plain
+  // response and once for performing Brotli compression. Since response.body is
+  // a ReadableStream and can only be read once under normal circumstances, we
+  // must call .clone on it so we can read it twice.
+  const responseClone = response.clone();
+
+  const cache = await openInjectorsCache();
+  const plainResponsePromise = cache
+    .put(cacheKey(url, key), response)
+    .then(() => {
+      console.log('> Plain response cached');
+    });
+
+  // Create an equivalent Response object that is already Brotli-compressed.
+  const compressed = await brotli.compress(
+    new Uint8Array(await responseClone.arrayBuffer()),
+    {
+      quality: 11,
+    }
+  );
+  const brotliResponse = new Response(compressed, {
+    headers: [
+      ...response.headers,
+      [HeaderKeys.CONTENT_ENCODING, ContentEncoding.BROTLI],
+    ],
+    encodeBody: 'manual',
+  });
+  const brotliResponsePromise = cache
+    .put(cacheKeyBrotli(url, key), brotliResponse)
+    .then(() => {
+      console.log('> Brotli response cached');
+    });
+
+  // Cache both the plain and the Brotli-compressed responses.
+  await Promise.all([plainResponsePromise, brotliResponsePromise]);
+}
+
+/**
+ * Enqueues a cache-save action for after the request, and respond with a clone
+ *
+ * @param extend - wrapper for the FetchEvent's waitUntil method.
+ * @param response - raw Response to save to cache.
+ * @param url - base file URL that has dynamic content injected into it.
+ * @param key - unique identifiers for this specific injection.
+ * @returns a clone of `response`.
+ */
+export function enqueueCacheAndClone(
```

Comment:
I think if multiple requests come in before the brotli compression is complete, we'll end up tasking multiple workers to compress the same contents. Can we build a mutex guard into the store?

Reply:
Unfortunately no, Cloudflare doesn't provide a global mutex primitive. Fortunately the result is idempotent, so at most we'll have a few hundred wasted CPU seconds :)